### PR TITLE
Remove timer on UndoRedoHelper and manually manage undo/redo snapshots a bit more

### LINKED ISSUE #3370: Don't create DispatcherTimers for TextBox undo/redo
It seems that each `TextBox` instance creates an `UndoRedoHelper` which in turn creates a `DispatcherTimer`. This means that we're creating a `DispatcherTimer` for every instance of `TextBox`, and this timer doesn't seem to get removed either.

I don't think a timer should be required for undo/redo: when I've done this sort of thing before the way it usually works is: each text change pushes to the undo stack. If the change is made below a certain time threshold then the undo stack entry is merged with the previous stack entry.

### PR DESCRIPTION
**Pull request by Deadpikle**
## What does the pull request do?

Improve the undo/redo system for `TextBox` by removing the need for `UndoRedoHelper` to use a 1-second timer to store things. (The timer was causing weird issues that were hard to reproduce on my end -- sometimes I couldn't undo or redo as I expected.) The `UndoRedoHelper` class was only used by `TextBox`, nor was it `public`, so this shouldn't have unintended side effects outside of `TextBox`. Importantly, this will make undo redo **consistent** because it no longer relies on a timer.

This also (in my opinion) improves undo/redo for `TextBox` by manually managing the stack a little bit more to be more intuitive: snapshot on space, snapshot in the middle of long words (every 7 changes; 7 was arbitrary), things like that.

I had trouble getting the initial state captured (e.g. from `` to `a`), so I added a flag to manually get that edge case, which I don't like; perhaps someone can see a better way to get that snapshot in. I'm also not sure this will work well for non-English languages, but I think (??) it should work OK since we snapshot in the middle of long words. **It would be helpful to me if someone could test this.**

I'm guessing this should have some sort of perf improvement, since we don't have a timer going off every 1 second for every textbox that's been created, but I don't know how to show this nicely in this PR with pretty graphs or similar.

By snapshotting before an undo is about to occur, we removed the need to snapshot after a text change (e.g. a snapshot -> handle text change -> another snapshot).

This also fixes a bug where the `UndoRedoHelper.Snapshot()` func to `_currentNode.Value.Equals(current)` did not properly call the `public bool Equals(UndoRedoState other)`. Not calling the right `Equals` func could cause an undo operation to snapshot the same text (but with a different `CaretIndex` as a different item to snapshot rather than a different one.

In summary, this PR:

* Removes need for undo redo helper snapshot timer that happened every 1s for every `TextBox` (a perf improvement at some level)
* Removes extra `Snapshot()` calls for undo/redo that are no longer necessary due to a snapshot that takes place before and undo operation
* Fixes `UndoRedoState.Equals` use
* Makes undo/redo a little nicer to use and much more consistent (because not relying on a timer)

### Before

(Note the weird undo to the middle of `hello`, probably caused by the timer)

before

### After

after

## What is the current behavior?

The `UndoRedoHelper` class uses a timer every 1 second to snapshot the text.

## What is the updated/expected behavior with this PR?

Got rid of the timer and manually managed things.

## How was the solution implemented (if it's not obvious)?

It's fairly straight forward. Nothing magical, really.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

Undo/redo functions differently now, but I think it's an improvement.

## Fixed issues
Fixes #3370

### CODE REVIEW

File: src/Avalonia.Controls/TextBox.cs [hunk 1]
```diff
@@ -446,6 +462,44 @@ public bool CanPaste
...
+        /// <summary>
+        /// Property for determining whether undo/redo is enabled
+        /// </summary>
+        public bool IsUndoEnabled
+        {
+            get { return GetValue(IsUndoEnabledProperty); }
+            set 
+            { 
+                SetValue(IsUndoEnabledProperty, value);
+                if (value == false)
```
Comment (maxkatz6) [inline src/Avalonia.Controls/TextBox.cs:474]: Property changed handling should be done in OnPropertyChanged method override or related observers. This code won't code if user will set property with SetValue(IsUndoEnabledProperty, value) directly.
For UndoLimit it's not that important, it's direct property that always will call setter, but makes sense for consistency I would say.

Comment (Deadpikle) [reply src/Avalonia.Controls/TextBox.cs]: Thanks. `OnPropertyChanged` now handles `IsUndoEnabledProperty`. I also fixed the bad getter/setter for `UndoLimit`. (This goes for the other PR, too -- I won't comment there so you don't get spammed.)

**Comment by Deadpikle** [comment]
Tried to run some benchmarks.

```csharp
[Benchmark]
public TextBox InitializeTextBox()
{
    using (UnitTestApplication.Start(TestServices.MockThreadingInterface))
    {
        return new TextBox();
    }
}
```

```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.1.410
  [Host]     : .NET Core 3.1.16 (CoreCLR 4.700.21.26205, CoreFX 4.700.21.26205), X64 RyuJIT
  DefaultJob : .NET Core 3.1.16 (CoreCLR 4.700.21.26205, CoreFX 4.700.21.26205), X64 RyuJIT
```

#### Before (master at beaf783b7)

|            Method |         Mean |        Error |       StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |-------------:|-------------:|-------------:|-------:|------:|------:|----------:|
| InitializeTextBox | 159,134.0 ns | 17,300.34 ns | 49,358.84 ns |      - |     - |     - |   16328 B |

#### With this change (the branch is a little old but oh well)

|            Method |         Mean |        Error |       StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |-------------:|-------------:|-------------:|-------:|------:|------:|----------:|
| InitializeTextBox | 125,612.0 ns | 10,671.53 ns | 31,465.26 ns |      - |     - |     - |   13656 B |

So, a small improvement, I suppose.

**Review by maxkatz6 — APPROVED** [review]

**Comment by maxkatz6** [comment]
Removing "backport candidate" just to be safe about it. Keeping in mind that this PR does not adds any functionality, but optimizes existing.

**Pull request merged**